### ISSUE TITLE
Optionally use nftables instead of iptables

### ISSUE THREAD

**Issue by vadorovsky** [cilium/cilium#4825]
iptables is used directly in several places inside `daemon/daemon.go`:

https://github.com/cilium/cilium/blob/master/daemon/daemon.go#L425
https://github.com/cilium/cilium/blob/master/daemon/daemon.go#L463
https://github.com/cilium/cilium/blob/master/daemon/daemon.go#L472-L478
https://github.com/cilium/cilium/blob/master/daemon/daemon.go#L483
https://github.com/cilium/cilium/blob/master/daemon/daemon.go#L547-L650

It would be nice to use nftables instead (if possible, if nftables is available).

There is a native Go library which we might use for that: https://github.com/google/nftables

To leave the choice to user, we can introduce a `--firewall-backend` option to `cilium-agent` and then resolve the firewall backend (nftables or iptables). Empty `--firewall-backend` will mean auto check for nftables with fallback to iptables.

**Comment by maci0** [comment]
This should be re-opened as Centos8 ships with nftables by default. iptables is a wrapper arount nft, iptables-legacy is not an option there and cilium tries to create incompatible rules.

**Comment by maci0** [comment]
In centos8 iptables is just a symlink to xtables-nft-multi.
debian buster does the same but has the possibility to use legacy iptables.
so either make use of nftables directly, or through the xtables-nft-multi iptables compatibility wrapper (needs iptables 1.8+ inside the image)

**Comment by maci0** [comment]
This seems to be addressed in https://github.com/cilium/cilium/commit/3a4fbd5fefec59ab50efc2c5bd271826e6f495ed#diff-ac50620a7dd6f7fef653345b64f848ab already.
The iptables-wrapper script there could also test if ip_tables/nftables is loaded in the kernel first.
if for example nftables is loaded but ip_tables isnt then the whole `iptables-legacy-save` can be skipped

**Comment by maci0** [comment]
another more fundamental question is why the `firewall` cni plugin isn't used. it supports multiple backends like `iptables` `nftables` `firewalld` https://github.com/containernetworking/plugins/tree/master/plugins/meta/firewall

**Comment by maci0** [comment]
@mskrocki for some reason this will cause the module to be loaded though. does not happen when the kube-proxy image performs the check.

so for now I had to rename the ip_tables.ko.xz file to something else. otherwise the module will get loaded as soon as I start cilium

**Comment by mskrocki** [comment]
Cilium container has all the modules loaded for ip_tables and nf_tables, it only controls the links of iptables cmd (same as kube-proxy). Where are the modules present/loaded, root or container namespace?

Additionally, did you run this new version of cilium on a clean system, or just upgraded cilium?
Thing is, if you had old version of cilium that just directly wrote to ip_tables (not nf_tables), the new version of cilium will see those written rules, and just still select ip_tables. To make sure the wrapper script works the ip_tables has to be flushed (if it is the nf_table-based system) before it starts.

**Comment by maci0** [comment]
Like I said, the kernel module is not even loaded before cilium starts. Therefore no legacy rules to flush as there are none. The moment cilium starts however it causes the module from the host systems /lib/modules/ to be loaded, even when it is blacklisted on the host's modprobe configuration.

**Comment by maci0** [comment]
```
[root@centos ~]# cat /etc/modprobe.d/iptables-blacklist.conf
blacklist ip_tables
install ip_tables /bin/false
[root@centos ~]#
```

```
[root@centos ~]# lsmod|grep 'ip_tables'
[root@centos ~]#
```


```
[root@centos ~]# iptables --version
iptables v1.8.2 (nf_tables)
```


```
[root@centos ~]# iptables -L |grep legacy
[root@centos ~]#
```


```
[root@centos ~]# iptables -L | grep kube
KUBE-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes service portals */
KUBE-EXTERNAL-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes externally-visible service portals */
KUBE-FORWARD  all  --  anywhere             anywhere             /* kubernetes forwarding rules */
KUBE-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes service portals */
KUBE-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes service portals */
DROP       all  --  anywhere             anywhere             /* kubernetes firewall for dropping marked packets */ mark match 0x8000/0x8000
REJECT     tcp  --  anywhere             anywhere             /* kube-system/hubble-ui-lb: has no endpoints */ ADDRTYPE match dst-type LOCAL tcp dpt:32718 reject-with icmp-port-unreachable
REJECT     tcp  --  anywhere             10.97.199.134        /* kube-system/hubble-ui-lb: has no endpoints */ tcp dpt:http reject-with icmp-port-unreachable
REJECT     tcp  --  anywhere             192.168.0.210        /* kube-system/hubble-ui-lb: has no endpoints */ tcp dpt:http reject-with icmp-port-unreachable
REJECT     udp  --  anywhere             10.96.0.10           /* kube-system/kube-dns:dns has no endpoints */ udp dpt:domain reject-with icmp-port-unreachable
REJECT     tcp  --  anywhere             10.96.0.10           /* kube-system/kube-dns:dns-tcp has no endpoints */ tcp dpt:domain reject-with icmp-port-unreachable
REJECT     tcp  --  anywhere             10.96.0.10           /* kube-system/kube-dns:metrics has no endpoints */ tcp dpt:9153 reject-with icmp-port-unreachable
ACCEPT     all  --  anywhere             anywhere             /* kubernetes forwarding rules */ mark match 0x4000/0x4000
ACCEPT     all  --  anywhere             anywhere             /* kubernetes forwarding conntrack pod source rule */ ctstate RELATED,ESTABLISHED
ACCEPT     all  --  anywhere             anywhere             /* kubernetes forwarding conntrack pod destination rule */ ctstate RELATED,ESTABLISHED
```

After the install

```
[root@centos ~]# helm upgrade -i cilium cilium/cilium --version 1.8.0-rc1 --namespace kube-system --set global.hubble.enabled=true --set global.hubble.relay.enabled=true --set global.hubble.ui.enabled=true --set global.hubble.cli.enabled=true --set global.hubble.listenAddress=
":4244" --set global.ipam.operator.clusterPoolIPv4PodCIDR="10.217.0.0/16"     --set config.ipam=kubernetes --set global.ipam.operator.mode=kubernetes --set global.hubble.ui.image.tag=latest
Release "cilium" does not exist. Installing it now.
NAME: cilium
LAST DEPLOYED: Mon Jun  1 21:56:19 2020
NAMESPACE: kube-system
STATUS: deployed
REVISION: 1
TEST SUITE: None
NOTES:
You have successfully installed Cilium with hubble-ui.

Your release version is 1.8.0-rc1.

For any further help, visit https://docs.cilium.io/en/v1.8/gettinghelp
[root@centos ~]#
```

```
[root@centos ~]# lsmod|grep 'ip_tables'
ip_tables              28672  4 iptable_filter,iptable_raw,iptable_nat,iptable_mangle
```

```
[root@centos ~]# iptables -L |grep legacy
# Warning: iptables-legacy tables present, use iptables-legacy to see them
```

As you can see kube-proxy did not cause the module load. But the moment cilium starts it gets loaded.
After cilium starts the kube api becomes also becomes inaccessible.

The only way to work around so far is to rename the ip_tables module.ko.xz file to something else

**Comment by mskrocki** [comment]
Could you try and run `iptables -L` inside one of the cilium pods?
The additional modules are loaded by cilium, which doesn't hurt, but probably could be conditioned in the code.
I suspect your system still has entries in the legacy iptables (created by previous version of cilium), which you should try and remove e.g. remove cilium from k8s, start cilium container directly with docker, from within that container clean the iptables.

**Comment by maci0** [comment]
Like I said. It's impossible there are legacy rules as there is no ip_tables kernel module loaded on the host.

```
[root@centos ~]# kubectl -n kube-system exec -ti pod/cilium-fnmqw -- iptables -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination
CILIUM_INPUT  all  --  anywhere             anywhere             /* cilium-feeder: CILIUM_INPUT */
KUBE-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes service portals */
KUBE-EXTERNAL-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes externally-visible service portals */
KUBE-FIREWALL  all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED,DNAT
ACCEPT     all  --  anywhere             anywhere
INPUT_direct  all  --  anywhere             anywhere
INPUT_ZONES  all  --  anywhere             anywhere
DROP       all  --  anywhere             anywhere             ctstate INVALID
REJECT     all  --  anywhere             anywhere             reject-with icmp-host-prohibited

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination
CILIUM_FORWARD  all  --  anywhere             anywhere             /* cilium-feeder: CILIUM_FORWARD */
KUBE-FORWARD  all  --  anywhere             anywhere             /* kubernetes forwarding rules */
KUBE-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes service portals */
ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED,DNAT
ACCEPT     all  --  anywhere             anywhere
FORWARD_direct  all  --  anywhere             anywhere
FORWARD_IN_ZONES  all  --  anywhere             anywhere
FORWARD_OUT_ZONES  all  --  anywhere             anywhere
DROP       all  --  anywhere             anywhere             ctstate INVALID
REJECT     all  --  anywhere             anywhere             reject-with icmp-host-prohibited

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
CILIUM_OUTPUT  all  --  anywhere             anywhere             /* cilium-feeder: CILIUM_OUTPUT */
KUBE-SERVICES  all  --  anywhere             anywhere             ctstate NEW /* kubernetes service portals */
KUBE-FIREWALL  all  --  anywhere             anywhere
ACCEPT     all  --  anywhere             anywhere
OUTPUT_direct  all  --  anywhere             anywhere

Chain INPUT_direct (1 references)
target     prot opt source               destination

Chain INPUT_ZONES (1 references)
target     prot opt source               destination
IN_public  all  --  anywhere             anywhere            [goto]
IN_public  all  --  anywhere             anywhere            [goto]

Chain FORWARD_direct (1 references)
target     prot opt source               destination

Chain FORWARD_IN_ZONES (1 references)
target     prot opt source               destination
FWDI_public  all  --  anywhere             anywhere            [goto]
FWDI_public  all  --  anywhere             anywhere            [goto]

Chain FORWARD_OUT_ZONES (1 references)
target     prot opt source               destination
FWDO_public  all  --  anywhere             anywhere            [goto]
FWDO_public  all  --  anywhere             anywhere            [goto]

Chain OUTPUT_direct (1 references)
target     prot opt source               destination

Chain IN_public (2 references)
target     prot opt source               destination
IN_public_pre  all  --  anywhere             anywhere
IN_public_log  all  --  anywhere             anywhere
IN_public_deny  all  --  anywhere             anywhere
IN_public_allow  all  --  anywhere             anywhere
IN_public_post  all  --  anywhere             anywhere
ACCEPT     icmp --  anywhere             anywhere

Chain IN_public_pre (1 references)
target     prot opt source               destination

Chain IN_public_log (1 references)
target     prot opt source               destination

Chain IN_public_deny (1 references)
target     prot opt source               destination

Chain IN_public_allow (1 references)
target     prot opt source               destination
ACCEPT     tcp  --  anywhere             anywhere             tcp dpt:ssh ctstate NEW,UNTRACKED
ACCEPT     tcp  --  anywhere             anywhere             tcp dpt:9090 ctstate NEW,UNTRACKED
ACCEPT     tcp  --  anywhere             anywhere             tcp dpt:6443 ctstate NEW,UNTRACKED
ACCEPT     tcp  --  anywhere             anywhere             tcp dpt:4244 ctstate NEW,UNTRACKED

Chain IN_public_post (1 references)
target     prot opt source               destination

Chain FWDI_public (2 references)
target     prot opt source               destination
FWDI_public_pre  all  --  anywhere             anywhere
FWDI_public_log  all  --  anywhere             anywhere
FWDI_public_deny  all  --  anywhere             anywhere
FWDI_public_allow  all  --  anywhere             anywhere
FWDI_public_post  all  --  anywhere             anywhere
ACCEPT     icmp --  anywhere             anywhere

Chain FWDI_public_pre (1 references)
target     prot opt source               destination

Chain FWDI_public_log (1 references)
target     prot opt source               destination

Chain FWDI_public_deny (1 references)
target     prot opt source               destination

Chain FWDI_public_allow (1 references)
target     prot opt source               destination

Chain FWDI_public_post (1 references)
target     prot opt source               destination

Chain FWDO_public (2 references)
target     prot opt source               destination
FWDO_public_pre  all  --  anywhere             anywhere
FWDO_public_log  all  --  anywhere             anywhere
FWDO_public_deny  all  --  anywhere             anywhere
FWDO_public_allow  all  --  anywhere             anywhere
FWDO_public_post  all  --  anywhere             anywhere

Chain FWDO_public_pre (1 references)
target     prot opt source               destination

Chain FWDO_public_log (1 references)
target     prot opt source               destination

Chain FWDO_public_deny (1 references)
target     prot opt source               destination

Chain FWDO_public_allow (1 references)
target     prot opt source               destination

Chain FWDO_public_post (1 references)
target     prot opt source               destination

Chain KUBE-FIREWALL (2 references)
target     prot opt source               destination
DROP       all  --  anywhere             anywhere             /* kubernetes firewall for dropping marked packets */ mark match 0x8000/0x8000

Chain KUBE-KUBELET-CANARY (0 references)
target     prot opt source               destination

Chain KUBE-PROXY-CANARY (0 references)
target     prot opt source               destination

Chain KUBE-EXTERNAL-SERVICES (1 references)
target     prot opt source               destination

Chain KUBE-SERVICES (3 references)
target     prot opt source               destination

Chain KUBE-FORWARD (1 references)
target     prot opt source               destination
DROP       all  --  anywhere             anywhere             ctstate INVALID
ACCEPT     all  --  anywhere             anywhere             /* kubernetes forwarding rules */ mark match 0x4000/0x4000
ACCEPT     all  --  anywhere             anywhere             /* kubernetes forwarding conntrack pod source rule */ ctstate RELATED,ESTABLISHED
ACCEPT     all  --  anywhere             anywhere             /* kubernetes forwarding conntrack pod destination rule */ ctstate RELATED,ESTABLISHED

Chain CILIUM_INPUT (1 references)
target     prot opt source               destination
ACCEPT     all  --  anywhere             anywhere             mark match 0x200/0xf00 /* cilium: ACCEPT for proxy traffic */

Chain CILIUM_OUTPUT (1 references)
target     prot opt source               destination
ACCEPT     all  --  anywhere             anywhere             mark match 0xa00/0xfffffeff /* cilium: ACCEPT for proxy return traffic */
MARK       all  --  anywhere             anywhere             mark match ! 0xe00/0xf00 mark match ! 0xd00/0xf00 mark match ! 0xa00/0xe00 /* cilium: host->any mark as from host */ MARK xset 0xc00/0xf00

Chain CILIUM_FORWARD (1 references)
target     prot opt source               destination
ACCEPT     all  --  anywhere             anywhere             /* cilium: any->cluster on cilium_host forward accept */
ACCEPT     all  --  anywhere             anywhere             /* cilium: cluster->any on cilium_host forward accept (nodeport) */
ACCEPT     all  --  anywhere             anywhere             /* cilium: cluster->any on lxc+ forward accept */
```

**Comment by maci0** [comment]
```
[root@centos 4.18.0-147.8.1.el8_1.x86_64]# kubectl -n kube-system exec -ti pod/cilium-fnmqw -- iptables-legacy -L
Chain INPUT (policy ACCEPT)
target     prot opt source               destination

Chain FORWARD (policy ACCEPT)
target     prot opt source               destination

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination
```

**Comment by maci0** [comment]
in any case. invoking `iptables` from within the container should not be allowed to load the kernel module, especially since it's blacklisted on the host

**Comment by mskrocki** [comment]
iptables modules are loaded by cilium: https://github.com/cilium/cilium/blob/master/pkg/datapath/iptables/iptables.go#L358
So, invoking iptables cmd doesn't load the modules.

Regardless, from the logs above it looks like the rules are created in the correct tables, with that, are you still experiencing connectivity problems?

**Comment by maci0** [comment]
I see. I think that warrants opening another issue.

I think connectivity issues were just co-incidental because of this one https://github.com/cilium/cilium/issues/11790

**Comment by mosfet1kg** [comment]
@maci0 Thank you for your kind comments.
I'm also in a trouble that may cause from iptable modules.
I found that a cluster ip for k8s api server doesn't work after cilium starts on RHEL8 env.
It's quite a time consuming and still not resolved.

I have a question to you and please answer me if possible.
I got a clue from your comments and then 
I've loaded following modules by using modprobe before starting cilium.
Then I cannot find the aforementioned problems and it seems to be resolved.
Does it affect to solve problem?
Thank you in advance.😂

    - ip_tables
    - iptable_raw
    - iptable_mangle
    - iptable_nat
    - iptable_filter
    - ip_tables
    - nf_nat_ipv4
    - nf_nat
    - nf_conntrack
    - nf_tables
    - nfnetlink

**Comment by flokli** [comment]
Is there any plans from Cilium to actually start supporting nftables directly if supported, rather than shelling out to different versions of the `iptables` binary (which this issue originally was opened for)?

It seems [some of the things Cilium does is not working in iptables-nft wrapper](https://github.com/NixOS/nixpkgs/issues/179741#issuecomment-1328402869), and it should be a good idea to properly support nftables anyways, not via some backwards-compat iptables wrapper script.

**Comment by joestringer** [comment]
@flokli unfortunately, depending on how other software on the system (eg kubelet) interacts with iptables/netfilter, Cilium needs to follow suit and use the same tooling. If there is a divergence, for instance if Cilium used nftables but other software used iptables, then this could provide misaligned rules to the kernel, leading to network traffic disruption. What makes this tricky is that Cilium containers ship their own ipt/nft tooling which we must make sure aligns with either the system or software used by other containers on the system. For now, we rely on shipping both iptables backends + the compatibility wrapper script (same as k8s) to select whether to use the legacy or nft backend to iptables. This mitigates most of the compatibility issues with other software.

My vote would be to push more on https://github.com/cilium/cilium/issues/12879 instead.

**Comment by flokli** [comment]
Yeah, that would also work, but considering that issue hasn't been touched for quite a while I'm not sure what will happen first. Do you have any plans to move one of these two tracks forward?

**Comment by joestringer** [comment]
At this point I believe that all nft compatibility issues are addressed. If you do not think so, please file a fresh issue calling this out and providing reproduction steps to demonstrate the compatibility issue.

If someone would like to work on direct nftables usage, then as long as the proposal satisfies the following constraints, then it may make sense in principle to accept such a change because it's unlikely that netfilter will completely disappear any time soon:
- As long as there are any users running iptables, there must still be a way to leverage the existing paths to install rules into iptables directly, and
- Hence the nuanced compatibility solution that I described in my last comment must also be retained in any such proposal, ie the new implementation must only be active iff the kubernetes canary rule exists in nftables and not iptables, and
- The way that the code is refactored to support the new approach is maintainable for existing logic paths that directly configure ipt/nft via CLI. This is to ensure that if we make subsequent changes to the rules, we can apply those changes consistently for all backends including the existing code, and
- There should be tests to demonstrate that the current approach that configures nftables will generate the exact same resulting rules as the new implementation (perhaps via golden tests or the control plane testing framework?).

If you have any additional compelling arguments for how such a proposal will improve Cilium's operations, performance or usability then I would welcome comments that describe these improved properties. This would help to motivate why someone may want to pick up this work.

I've dropped the `pinned` label since it's not clear to me that there is sufficient community interest in driving this proposal forward. If someone finds this issue in future and is interested in contributing to it, then you are welcome to make a proposal based on the constraints outlined above. Feel free to ask any questions if the guidance above is unclear. The next concrete step forward would be a more comprehensive writeup about how this will be solved, or a PR that proposes to make the actual changes to the tree.
